File: DSA/src/go/validate_binary_search_tree.go
package main

import "math"

func isValidBST(root *TreeNode) bool {
	var helper func(node *TreeNode, upperLimit, lowerLimit int) bool
	helper = func(node *TreeNode, upperLimit, lowerLimit int) bool {
		if node == nil {
			return true
		}
		return lowerLimit <= node.Val && node.Val <= upperLimit && helper(node.Left, node.Val-1, lowerLimit) &&
			helper(node.Right, upperLimit, node.Val+1)
	}
	return helper(root, math.MaxInt, math.MinInt)
}

File: docs/DSA/src/py/validate_binary_search_tree.py
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def _helper(self, node, upper_bound, lower_bound):
        if not node:
            return True

        return lower_bound <= node.val <= upper_bound and \
            self._helper(node.left, node.val-1, lower_bound) and \
            self._helper(node.right, upper_bound, node.val+1)

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self._helper(root, float('inf'), float('-inf'))
